make-kamikaze-2.1.sh

```diff
@@ -159,11 +157,10 @@ create_user() {
 
 install_redeem() {
 	echo "**install_redeem**"
-	cd /usr/src/
-	if [ ! -d "redeem" ]; then
-		git clone --no-single-branch --depth 1 https://github.com/intelligent-agent/redeem
+	if [ ! -d "/usr/src/redeem" ]; then
+		git clone --no-single-branch --depth 1 https://github.com/intelligent-agent/redeem /usr/src/redeem
```

Comment:
Use of hard hard-coded addresses makes it much more difficult to use the code in another context, for example, building a development sandbox.  Thus, rather than having "/usr/src/redeem" we should use the autoconf style of having that address stored in a script variable which was defined at the top of the script.

Reply:
There should be the `$WD` variable set at the top of the script already. And it should be re-used here.

Reply:
the popd/pushd combo only makes so much sense at times, having the $WD variable set and re-used does make it a lot easier to manage path changes.

Reply:
I'd go one step farther and use $WD in the definition of a variable for each source tree. It is highly likely, for example, that I would want to use the standard Octoprint, etc. but have my experimental redeem tree elsewhere.

Reply:
I nuked $WD with prejudice after finding that the script was getting to the end and failing to delete a random file that was supposed to be in the current directory... which means at some point it cd'd to a different directory and didn't go back. And since it had root privileges, it happily continued on and wrecked the image for an unknowable amount of time before hitting something that failed and stopped it. I wasn't willing to debug that or flood the file with `cd $WD` everywhere possible, so I used absolute paths as much as possible and used pushd/popd when I couldn't.

A few more general remarks:
- Why would you be rebuilding an Umikaze image to make a development sandbox? If the whole point of the sandbox is that you're going to be changing things quite rapidly and getting your hands dirty, why would you start by running a > 1 hour image generation script instead of spending <1 minute to delete/remove the initial repo from an already-finished image?
- Why would you want to change the repo location when git means you can mess around with the original and instantly restore it to its original state with one command whenever you want?

Reply:
To be clear, I don't really have a problem with trying to further parameterize this, but I think a better option would probably be to make each `install_x` function take a target path as a parameter and define all the paths separately at the top of the file.

Taking it a step further, suppose we moved each `install_x` function out into its own script? That would mean the main script would just have stuff like `( cd ${REDEEM_INSTALL_DIR} && ${SCRIPT_PATH}/modules/install_redeem.sh )` which means it's impossible for the installer function to get us into a bad directory?

Reply:
That sounds like a reasonably good idea. But I would argue that setting the REDEEM_INSTALL_DIR folder should still be done in the main script. That way we have one file with all the variables defined to modify when building a new image, not a slew of them.

Reply:
I agree, although I'd argue that these variable should be moved out to a separate "config file" even if that file is also a bash script that the main script sources. In the long run, if we're going to somehow use this to automate CI builds of Umikaze images, we'll want the config to be easy to generate.

Reply:
Yeah, we can do that. I'm just unsure whether that refactor should happen before or after your changes are merged?